src/js/mixins/evented.js

```diff
@@ -0,0 +1,385 @@
+ * the type of target.
+ *
+ * @private
+ * @param  {Element|Object} target
+ *         A DOM node or evented object.
+ *
+ * @param  {string|Array} type
+ *         One or more event type(s).
+ *
+ * @param  {Function} listener
+ *         A listener function.
+ *
+ * @param  {string} [method="on"]
+ *         The event binding method to use.
+ */
+const listen = (target, type, listener, method = 'on') => {
+  validateTarget(target);
+
+  if (target.nodeName) {
+    Events[method](target, type, listener);
+  } else {
+    target[method](type, listener);
+  }
+};
+
+/**
+ * Methods that can be mixed-in with any object to provide event capabilities.
+ *
+ * @name mixins/evented
+ * @type {Object}
+ */
+const mixin = {
+
+  /**
+   * Add a listener to an event (or events) on this object or another evented
+   * object.
+   *
+   * @param  {string|Array|Element|Object} targetOrType
+   *         If this is a string or array, it represents an event type(s) and
+   *         the listener will listen for events on this object.
+   *
+   *         Another evented object can be passed here instead, which will
+   *         cause the listener to listen for events on THAT object.
+   *
+   *         In either case, the listener's `this` value will be bound to
+   *         this object.
+   *
+   * @param  {string|Array|Function} typeOrListener
+   *         If the first argument was a string or array, this should be the
+   *         listener function. Otherwise, this is a string or array of event
+   *         type(s).
+   *
+   * @param  {Function} [listener]
+   *         If the first argument was another evented object, this will be
+   *         the listener function.
+   *
+   * @return {Object}
+   *         Returns this object.
+   */
+  on(...args) {
+    const {isTargetingSelf, target, type, listener} = normalizeListenArgs(this, args);
+
+    listen(target, type, listener);
+
+    // If this object is listening to another evented object.
+    if (!isTargetingSelf) {
+
+      // If this object is disposed, remove the listener.
+      const removeListenerOnDispose = () => this.off(target, type, listener);
+
+      // Use the same function ID as the listener so we can remove it later it
+      // using the ID of the original listener.
+      removeListenerOnDispose.guid = listener.guid;
+
+      // Add a listener to the target's dispose event as well. This ensures
+      // that if the target is disposed BEFORE this object, we remove the
+      // removal listener that was just added. Otherwise, we create a memory leak.
+      const removeRemoverOnTargetDispose = () => this.off('dispose', removeListenerOnDispose);
+
+      // Use the same function ID as the listener so we can remove it later
+      // it using the ID of the original listener.
+      removeRemoverOnTargetDispose.guid = listener.guid;
+
+      listen(this, 'dispose', removeListenerOnDispose);
+      listen(target, 'dispose', removeRemoverOnTargetDispose);
+    }
+
+    return this;
```

Comment:
Will remove it now that we seem to have decided to remove method chaining across the board.